pinot-common/src/main/java/org/apache/pinot/common/failuredetector/BaseExponentialBackoffRetryFailureDetector.java

```diff
@@ -88,21 +102,28 @@ public void start() {
             LOGGER.info("Server: {} has been marked healthy, skipping the retry", instanceId);
             continue;
           }
-          if (retryInfo._numRetires == _maxRetries) {
+          if (retryInfo._numRetries == _maxRetries) {
             LOGGER.warn("Unhealthy server: {} already reaches the max retries: {}, do not retry again and treat it "
                 + "as healthy so that the listeners do not lose track of the server", instanceId, _maxRetries);
             markServerHealthy(instanceId);
             continue;
           }
           LOGGER.info("Retry unhealthy server: {}", instanceId);
-          for (Listener listener : _listeners) {
-            listener.retryUnhealthyServer(instanceId, this);
+          boolean recovered = true;
+          for (Function<String, Boolean> unhealthyServerRetrier : _unhealthyServerRetriers) {
+            if (!unhealthyServerRetrier.apply(instanceId)) {
+              recovered = false;
```

Comment:
I was just going to comment about this section when it was written the other way around, but it looks like you just changed it!

Reply:
Yeah with the previous implementation, it was possible for multiple listeners to mark the same server as healthy on a call to `retryUnhealthyServer`. I've updated the implementation so that the failure detector itself marks the server as healthy if all the listeners / retriers report that the server is healthy again (to be safe).